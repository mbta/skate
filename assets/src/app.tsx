// We need to import the CSS so that webpack will load it.
// The MiniCssExtractPlugin is used to separate it out into
// its own CSS file.
declare function require(name: string): string
// tslint:disable-next-line
require("../css/app.scss")

// webpack automatically bundles all modules in your
// entry points. Those entry points can be configured
// in "webpack.config.js".
//
// Import dependencies
//
import "@babel/polyfill"
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import "leaflet/dist/leaflet.css"
import "phoenix_html"
import * as React from "react"
import ReactDOM from "react-dom"
import AppStateWrapper from "./components/appStateWrapper"

if (window.FS && window.username) {
  window.FS.identify(window.username, { displayName: window.username })
}

// Import local files
//
// Local files can be imported directly using relative paths, for example:
// import socket from "./socket"

ReactDOM.render(<AppStateWrapper />, document.getElementById("app"))
