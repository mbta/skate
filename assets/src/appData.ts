const appData = (): DOMStringMap | undefined => {
  const dataEl = document.getElementById("app")

  if (!dataEl) {
    return undefined
  }

  return dataEl.dataset
}

export default appData
