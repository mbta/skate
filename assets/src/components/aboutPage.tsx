import React, { ReactElement } from "react"

const AboutPage = (): ReactElement<HTMLDivElement> => (
  <div className="c-page c-page--about">
    <div className="c-page__container">
      <h1 className="c-page__title">About Skate</h1>
      <p>
        Skate is an active work-in-progress. In the coming months, you’ll notice
        new tools like shuttle tracking. You can also expect changes and
        refinements to the route ladders as we continue building.
      </p>
      <div className="c-page__section">
        <h2 className="c-page__header">Device Support</h2>
        <p>Need tablet support? Call the MBTA Help Desk at x5761.</p>
      </div>
      <div className="c-page__section">
        <h2 className="c-page__header">Frequently Asked Questions</h2>
        <p>
          Have questions about using Skate? We’re making updates all the time.
          Read the latest news and refresh your training by visiting our FAQ.
        </p>
        <p>
          <a
            className="c-page__link"
            href="https://docs.google.com/document/d/e/2PACX-1vQVzXUcQWg-nKCkb8CyModNUCXDHHRxgqPdtlkm1GQOK9aCrGMSJAoC2J8JuuGFPGnPkcetrZEk6ERH/pub"
            target="_blank"
          >
            FAQ
          </a>
        </p>
      </div>
      <div className="c-page__section">
        <h2 className="c-page__header">Feedback</h2>
        <p>
          Have feedback, a suggestion, or need other support related to Skate?
          Use our feedback form and someone on our team will be in touch with
          you.
        </p>
        <p>
          <a
            className="c-page__link"
            href="https://form.asana.com?hash=e0ca8ee9a5fbde7691ba0ff428c826eba5a9700b99249e26124dd33d13b511ca&id=1155343357379913"
            target="_blank"
          >
            Give Feedback
          </a>
        </p>
      </div>
    </div>
  </div>
)

export default AboutPage
