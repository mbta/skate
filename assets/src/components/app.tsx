import React, { ReactElement, useContext } from "react"
import { BrowserRouter, Route as BrowserRoute } from "react-router-dom"
import { SocketContext } from "../contexts/socketContext"
import { StateDispatchContext } from "../contexts/stateDispatchContext"
import useAppcues from "../hooks/useAppcues"
import { ConnectionStatus } from "../hooks/useSocket"
import DataStatusBanner from "./dataStatusBanner"
import DisconnectedModal from "./disconnectedModal"
import LadderPage from "./ladderPage"
import SearchPage from "./searchPage"
import SettingsPage from "./settingsPage"
import ShuttleMapPage from "./shuttleMapPage"
import TabBar from "./tabBar"
import { Notifications } from "./notifications"

const AppRoutes = () => {
  useAppcues()

  const [{ pickerContainerIsVisible }] = useContext(StateDispatchContext)
  const { connectionStatus } = useContext(SocketContext)

  return (
    <div className="m-app">
      <div className="m-app__banner">
        <DataStatusBanner />
      </div>
      <div className="m-app__main">
        <Notifications />
        <TabBar pickerContainerIsVisible={pickerContainerIsVisible} />
        <BrowserRoute exact={true} path="/" component={LadderPage} />
        <BrowserRoute
          exact={true}
          path="/shuttle-map"
          component={ShuttleMapPage}
        />
        <BrowserRoute exact={true} path="/settings" component={SettingsPage} />
        <BrowserRoute exact={true} path="/search" component={SearchPage} />
        {connectionStatus === ConnectionStatus.Disconnected ? (
          <DisconnectedModal />
        ) : null}
      </div>
    </div>
  )
}

const App = (): ReactElement<HTMLDivElement> => {
  return (
    <BrowserRouter>
      <AppRoutes />
    </BrowserRouter>
  )
}

export default App
