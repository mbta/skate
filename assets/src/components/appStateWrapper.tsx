import React from "react"
import { StateDispatchProvider } from "../contexts/stateDispatchContext"
import { TripsByIdProvider } from "../contexts/tripsByIdContext"
import { VehiclesByRouteIdProvider } from "../contexts/vehiclesByRouteIdContext"
import usePersistedStateReducer from "../hooks/usePersistedStateReducer"
import useSocket from "../hooks/useSocket"
import useTrips from "../hooks/useTrips"
import useVehicles from "../hooks/useVehicles"
import { TripsById } from "../schedule.d"
import { initialState, reducer } from "../state"
import App from "./app"

const AppStateWrapper = (): JSX.Element => {
  const [state, dispatch] = usePersistedStateReducer(reducer, initialState)
  const { selectedRouteIds } = state

  const tripsById: TripsById = useTrips(selectedRouteIds)
  const socket = useSocket()
  const vehiclesByRouteId = useVehicles(socket, selectedRouteIds)

  return (
    <StateDispatchProvider state={state} dispatch={dispatch}>
      <TripsByIdProvider tripsById={tripsById}>
        <VehiclesByRouteIdProvider vehiclesByRouteId={vehiclesByRouteId}>
          <App />
        </VehiclesByRouteIdProvider>
      </TripsByIdProvider>
    </StateDispatchProvider>
  )
}

export default AppStateWrapper
