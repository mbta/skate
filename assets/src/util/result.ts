/**
 * The discriminator value used to determine if a {@linkcode Result<T, E>} is
 * {@linkcode Ok<T>} or {@linkcode Err<E>}.
 */
enum ResultTag {
  Ok,
  Err,
}

/**
 * The key of the {@linkcode Result<T, E>} discriminator value.
 *
 * This is not exported so that usages of the objects cannot be
 * accessed or modified manually outside of this module.
 */
const tag = Symbol("result-enum-tag")

/** The {@linkcode Ok<T>} variant of {@linkcode Result<T, E>} */
export type Ok<T> = { [tag]: ResultTag.Ok; ok: T }

/** The {@linkcode Err<E>} variant of {@linkcode Result<T, E>} */
export type Err<E> = { [tag]: ResultTag.Err; err: E }

/**
 * Type which may contain an {@linkcode Ok<T>} state, or a {@linkcode Err<E>} state
 *
 * useful for requiring a caller of a function to handle the Error state of a return value.
 */
export type Result<T, E> = Ok<T> | Err<E>

/** Determines if a {@linkcode Result<T, E>} is a {@linkcode Ok<T>} */
export const isOk = <T, E>(r: Result<T, E>): r is Ok<T> =>
  tag in r && r[tag] === ResultTag.Ok

/** Determines if a {@linkcode Result<T, E>} is a {@linkcode Err<E>} */
export const isErr = <T, E>(r: Result<T, E>): r is Err<E> =>
  tag in r && r[tag] === ResultTag.Err

/** Constructs a {@linkcode Ok<T>} with {@linkcode v} as the value */
export const Ok = <T>(v: T): Ok<T> => ({
  [tag]: ResultTag.Ok,
  ok: v,
})

/** Constructs a {@linkcode Err<E>} with {@linkcode v} as the value */
export const Err = <E>(v: E): Err<E> => ({
  [tag]: ResultTag.Err,
  err: v,
})

/**
 * If {@linkcode r} is {@linkcode Ok<T>}, it applies {@linkcode fn} to
 * the value contained in {@linkcode r}. Otherwise returns {@linkcode r}.
 */
export const map = <T, U, E>(r: Result<T, E>, fn: (v: T) => U): Result<U, E> =>
  isOk(r) ? Ok(fn(r.ok)) : r
