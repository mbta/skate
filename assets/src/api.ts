import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
