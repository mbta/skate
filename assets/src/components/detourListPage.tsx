import React, { useCallback, useState } from "react"
import { DetoursTable } from "./detoursTable"
import userInTestGroup, { TestGroups } from "../userInTestGroup"
import { Button } from "react-bootstrap"
import { PlusSquare } from "../helpers/bsIcons"
import { DetourModal } from "./detours/detourModal"
import { fetchDetours } from "../api"
import { useApiCall } from "../hooks/useApiCall"
import { isOk } from "../util/result"

export const DetourListPage = () => {
  const [showDetourModal, setShowDetourModal] = useState(false)

  const { result } = useApiCall({
    apiCall: useCallback(async () => fetchDetours(), []),
  })
  const detours = result && isOk(result) && result.ok

  return (
    <div className="c-detour-list-page h-100 overflow-y-auto p-0 p-md-4 bg-white">
      {userInTestGroup(TestGroups.DetoursPilot) && (
        <Button
          className="c-detour-list-page__button icon-link fw-light px-3 py-2 u-hide-for-mobile"
          onClick={() => setShowDetourModal(true)}
        >
          <PlusSquare />
          <span className="c-detour-list-page__button-text">Add detour</span>
        </Button>
      )}
      {detours && (
        <>
          {detours.active && <DetoursTable data={detours.active} />}
          {detours.draft && <DetoursTable data={detours.draft} />}
          {detours.past && <DetoursTable data={detours.past} />}
        </>
      )}

      {showDetourModal && (
        <DetourModal
          onClose={() => setShowDetourModal(false)}
          show
          originalRoute={{}}
        />
      )}
    </div>
  )
}
