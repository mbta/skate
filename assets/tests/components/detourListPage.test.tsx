import { describe, test, expect, jest, beforeEach } from "@jest/globals"
import React from "react"
import { DetourListPage } from "../../src/components/detourListPage"
import { fetchDetours } from "../../src/api"
import { neverPromise } from "../testHelpers/mockHelpers"
import { Ok } from "../../src/util/result"
import { render, screen } from "@testing-library/react"

jest.useFakeTimers().setSystemTime(new Date("2024-08-29T20:00:00"))

jest.mock("../../src/api")

beforeEach(() => {
  jest.mocked(fetchDetours).mockReturnValue(neverPromise())
})

describe("DetourListPage", () => {
  test("renders detour list page", async () => {
    jest.mocked(fetchDetours).mockResolvedValue(
      Ok({
        active: [
          {
            id: 1,
            route: "1",
            direction: "Inbound",
            name: "Headsign A",
            intersection: "Street A & Avenue B",
            updatedAt: 1724866392,
          },
          {
            id: 8,
            route: "2",
            direction: "Outbound",
            name: "Headsign B",
            intersection: "Street C & Avenue D",
            updatedAt: 1724856392,
          },
        ],
        draft: undefined,
        past: [
          {
            id: 10,
            route: "1",
            direction: "Inbound",
            name: "Headsign A",
            intersection: "Street E & Avenue F",
            updatedAt: 1724866392,
          },
          {
            id: 7,
            route: "1",
            direction: "Outbound",
            name: "Headsign Z",
            intersection: "Street C & Avenue D",
            updatedAt: 1724866392,
          },
        ],
      })
    )

    const { baseElement } = render(<DetourListPage />)

    await screen.findByText("Headsign Z")

    expect(baseElement).toMatchSnapshot()
  })
})
