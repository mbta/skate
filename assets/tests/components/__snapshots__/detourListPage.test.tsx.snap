// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetourListPage renders detour list page 1`] = `
<body>
  <div>
    <div
      class="c-detour-list-page h-100 overflow-y-auto p-0 p-md-4 bg-white"
    >
      <table
        class="c-detours-table table table-hover"
      >
        <thead
          class="u-hide-for-mobile"
        >
          <tr>
            <th
              class="px-3 py-4"
            >
              Route and direction
            </th>
            <th
              class="px-3 py-4 u-hide-for-mobile"
            >
              Starting Intersection
            </th>
            <th
              class="px-3 py-4 u-hide-for-mobile"
            >
              On detour since
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="align-middle p-3"
            >
              <div
                class="d-flex"
              >
                <div
                  class="c-route-pill c-route-pill--bus"
                >
                  1
                </div>
                <div
                  class="c-detours-table__route-info-text d-inline-block"
                >
                  <div
                    class="pb-1 fs-4 fw-semibold"
                  >
                    Headsign A
                  </div>
                  <div
                    class="c-detours-table__route-info-direction fs-6"
                  >
                    Inbound
                  </div>
                </div>
              </div>
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              Street A & Avenue B
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              26 hours ago
            </td>
          </tr>
          <tr>
            <td
              class="align-middle p-3"
            >
              <div
                class="d-flex"
              >
                <div
                  class="c-route-pill c-route-pill--bus"
                >
                  2
                </div>
                <div
                  class="c-detours-table__route-info-text d-inline-block"
                >
                  <div
                    class="pb-1 fs-4 fw-semibold"
                  >
                    Headsign B
                  </div>
                  <div
                    class="c-detours-table__route-info-direction fs-6"
                  >
                    Outbound
                  </div>
                </div>
              </div>
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              Street C & Avenue D
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              29 hours ago
            </td>
          </tr>
        </tbody>
      </table>
      <table
        class="c-detours-table table table-hover"
      >
        <thead
          class="u-hide-for-mobile"
        >
          <tr>
            <th
              class="px-3 py-4"
            >
              Route and direction
            </th>
            <th
              class="px-3 py-4 u-hide-for-mobile"
            >
              Starting Intersection
            </th>
            <th
              class="px-3 py-4 u-hide-for-mobile"
            >
              On detour since
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="align-middle p-3"
            >
              <div
                class="d-flex"
              >
                <div
                  class="c-route-pill c-route-pill--bus"
                >
                  1
                </div>
                <div
                  class="c-detours-table__route-info-text d-inline-block"
                >
                  <div
                    class="pb-1 fs-4 fw-semibold"
                  >
                    Headsign A
                  </div>
                  <div
                    class="c-detours-table__route-info-direction fs-6"
                  >
                    Inbound
                  </div>
                </div>
              </div>
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              Street E & Avenue F
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              26 hours ago
            </td>
          </tr>
          <tr>
            <td
              class="align-middle p-3"
            >
              <div
                class="d-flex"
              >
                <div
                  class="c-route-pill c-route-pill--bus"
                >
                  1
                </div>
                <div
                  class="c-detours-table__route-info-text d-inline-block"
                >
                  <div
                    class="pb-1 fs-4 fw-semibold"
                  >
                    Headsign Z
                  </div>
                  <div
                    class="c-detours-table__route-info-direction fs-6"
                  >
                    Outbound
                  </div>
                </div>
              </div>
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              Street C & Avenue D
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              26 hours ago
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
